src/core/init/duration_updater.ts

```diff
@@ -40,27 +62,139 @@ export default function DurationUpdater(
   manifest : Manifest,
   mediaSource : MediaSource
 ) : Observable<never> {
-  return fromEvent(manifest, "manifestUpdate").pipe(
-    startWith(null),
-    tap(() => {
-      const maximumPosition = manifest.getMaximumPosition();
+  return observableDefer(() => {
+    let lastSetDuration: number | undefined;
+    return setMediaSourceDuration(mediaSource, manifest, undefined).pipe(
+      mergeMap((firstSetDuration) => {
+        // only update `lastSetDuration` if the MediaSource's duration has been
+        // updated.
+        if (firstSetDuration !== null) {
+          lastSetDuration = firstSetDuration;
+        }
 
+        return fromEvent(manifest, "manifestUpdate").pipe(
+          switchMap(() => setMediaSourceDuration(mediaSource, manifest, lastSetDuration)),
+          tap((setDuration) => {
+            if (setDuration !== null) {
+              lastSetDuration = setDuration;
```

Comment:
The variable `setDuration` reminds me of a function that `set a duration`, does the name is appropriate?

Reply:
it's more "a set duration" in past form,  like "an updated duration", not the infinitive like "updateDuration".

But yes it can be confusing. I'll try to update it to something more understandable

Reply:
Renamed to `lastDurationUpdate`/`initialDurationUpdate`/`durationUpdate`